087-external-certificate-manager.md

```diff
@@ -0,0 +1,241 @@
+
+During a reconciliation Strimzi will check the hash of the certificate stored in the user's CA public cert Secret to see if an update is needed.
+If the certificate has changed Strimzi will rename the existing cert in the `<CLUSTER_NAME>-cluster-ca-cert` Secret to `ca-YYYY-MM-DDTHH-MM-SSZ.crt` and copy over the new certificate.
+It will also update the `strimzi.io/ca-cert-hash` and increment the `strimzi.io/cluster-ca-key-generation` and `strimzi.io/cluster-ca-cert-generation` annotations.
+
+#### Issuing end-entity certificates
+
+If the user has enabled this feature, when Strimzi needs to issue a certificate, instead of using the existing internal mechanism it will create a `Certificate` custom resource.
+Strimzi will specify the required CN/SANS in the `Certificate` resource for the end-entity certificate.
+Strimzi will specify the Secret for the certificate to be stored in as the existing name for the certificate Secret suffixed with `-cm`, for example `<CLUSTER_NAME>-cluster-operator-certs-cm`.
+Strimzi will wait for the usual operation timeout during the reconciliation loop for the `Certificate` status to indicate that the certificate has been issued before continuing.
+When issuing cluster certificates (e.g for each Kafka pod etc), once the certificate has been issued, Strimzi will copy the certificate across from the cert-manager provided Secret into its own existing Secret.
+Strimzi will annotate the Secret it manages with:
+* `strimzi.io/server-cert-hash` annotation with the value being the hash of the certificate in the cert-manager Secret.
+* `strimzi.io/cluster-ca-cert-generation` annotation with the value matching the current value on the `<CLUSTER_NAME>-cluster-ca-cert` Secret.
+
+Similar to today, Strimzi will also add both these annotations to the pod mounting the Secret.
+
+#### Tracking changes to cluster end-entity certificates
+
+Cert-manager will be responsible for renewing all end-entity certificates.
+When a certificate is renewed cert-manager will update the related Secret.
+During a reconciliation Strimzi will check the hash of the certificate stored in cert-manager Secrets to see if an update is needed.
+
+When cert-manager renews an end-entity certificate Strimzi will copy over the new certificate into its own Secret.
+Strimzi will update the `strimzi.io/server-cert-hash` annotation to match the new certificate.
+Strimzi will update the `strimzi.io/cluster-ca-cert-generation` annotation to match the annotation on the `<CLUSTER_NAME>-cluster-ca-cert` Secret.
+
+#### Handling CA key replacements
+
+During a CA key replacement the user must pause Strimzi reconciliation until the following are complete:
+* A new CA key has been generated
+* A new CA public certificate has been issued
+* The Secret provided to Strimzi containing the CA public certificate has been updated
+* Cert-manager has issued new end-entity certificates using the new CA public certificate
```

Comment:
Can you elaborate on why this is needed? And maybe also on what does happen here on the Cert Manager side? I assume you describe here a new CA used for existing issuer? 

A particularly interesting points:
* How will the user know that `Cert-manager has issued new end-entity certificates using the new CA public certificate`?
* Will the user include only the new CA in the trusted CAs? Or both the new and old and have the user delete the old CA later? I assume the second is correct, or?

----

Also, what about migration to a different Issuer? Should we have a chapter for that  as well?

Reply:
The flow would actually work without the last bullet point so I'll remove that

I'll add some more details on the Cert Manager side and look at what happens when migrating to a different Issuer.

On this point:
> Will the user include only the new CA in the trusted CAs? Or both the new and old and have the user delete the old CA later? I assume the second is correct, or?

Since Strimzi is managing copying the CA cert over into the <CLUSTER_NAME>-cluster-ca-cert Secret the user just needs to provide the new one. Strimzi will hold onto the old one until it has seen new end-entity certificates have been issued and then remove it as described in the section that steps through what Strimzi does during the next reconciliation.